src/server/coprocessor/endpoint.rs

```diff
@@ -981,4 +1218,141 @@ mod tests {
+        (8,String::from("data8"),Datum::Bytes(b"name:8".to_vec()), Datum::I64(2)),
+        (9,String::from("data9"),Datum::Bytes(b"name:9".to_vec()), Datum::I64(2)),
+        ];
+
+        let exp = vec![
+        (9,String::from("data9"),Datum::Bytes(b"name:9".to_vec()), Datum::I64(2)),
+        (8,String::from("data8"),Datum::Bytes(b"name:8".to_vec()), Datum::I64(2)),
+        (7,String::from("data7"),Datum::Bytes(b"name:7".to_vec()), Datum::I64(2)),
+        (3,String::from("data3"),Datum::Bytes(b"name:3".to_vec()), Datum::I64(1)),
+        (4,String::from("data4"),Datum::Bytes(b"name:3".to_vec()), Datum::I64(2)),
+        ];
+
+        for (handle, data, name, count) in test_data {
+            let cur_key: Vec<Datum> = vec![name, count];
+            let row_data = RowColsDict::new(HashMap::new(), data.into_bytes());
+            topn_heap.try_add_row(handle as i64,
+                             order_cols.clone(),
+                             ctx.clone(),
+                             row_data,
+                             cur_key)
+                .unwrap();
+        }
+        let result = topn_heap.into_sorted_vec().unwrap();
+        assert_eq!(result.len(), exp.len());
+        for (row, (handle, _, name, count)) in result.iter().zip(exp) {
+            let exp_keys: Vec<Datum> = vec![name, count];
+            assert_eq!(row.handle, handle);
+            assert_eq!(row.key, exp_keys);
+        }
+    }
+
+    #[test]
+    fn test_topn_heap_with_cmp_error() {
+        let mut order_cols = Vec::new();
+        order_cols.push(new_order_by(0, true));
+        order_cols.push(new_order_by(1, false));
+        let order_cols = Rc::new(order_cols);
+        let ctx = Rc::new(EvalContext::default());
+        let mut topn_heap = TopNHeap::new(5).unwrap();
+
+        let std_key: Vec<Datum> = vec![Datum::Bytes(b"aaa".to_vec()), Datum::I64(2)];
+        let row_data = RowColsDict::new(HashMap::new(), b"name:1".to_vec());
+        topn_heap.try_add_row(0 as i64, order_cols.clone(), ctx.clone(), row_data, std_key)
+            .unwrap();
+
+        let std_key2: Vec<Datum> = vec![Datum::Bytes(b"aaa".to_vec()), Datum::I64(3)];
+        let row_data2 = RowColsDict::new(HashMap::new(), b"name:2".to_vec());
+        topn_heap.try_add_row(0 as i64,
+                         order_cols.clone(),
+                         ctx.clone(),
+                         row_data2,
+                         std_key2)
+            .unwrap();
+
+        let bad_key1: Vec<Datum> = vec![Datum::I64(2), Datum::Bytes(b"aaa".to_vec())];
+        let row_data3 = RowColsDict::new(HashMap::new(), b"name:3".to_vec());
+
+        assert!(topn_heap.try_add_row(0 as i64,
+                         order_cols.clone(),
+                         ctx.clone(),
+                         row_data3,
+                         bad_key1)
+            .is_err());
+
+        assert!(topn_heap.into_sorted_vec().is_err());
+    }
+
+    #[test]
+    fn test_topn_heap_with_few_data() {
```

Comment:
`test_topn_heap` with more test data than limit which would make the heap full, while `test_topn_heap_with_few_data`'s number of test data  is less than limit.